### PR TITLE
Strip ANSI escape codes in TAP test titles

### PR DESCRIPTION
**Pull request by sotojuan**
Fixes issue mentioned by @xjamundx in #723.

### CODE REVIEW

File: lib/reporters/tap.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 'use strict';
 var format = require('util').format;
+var stripANSI = require('strip-ansi');
```
Comment (novemberborn) [inline lib/reporters/tap.js]: Human linting nitpick, I'd prefer `stripAnsi` actually 😉

File: lib/reporters/tap.js [hunk 2]
```diff
@@ -36,10 +37,12 @@ TapReporter.prototype.test = function (test) {
 		directive = '# SKIP';
 	}
 
+	var strippedTitle = stripANSI(test.title);
```
Comment (novemberborn) [inline lib/reporters/tap.js]: `title` should suffice.

**Comment by novemberborn** [comment]
Just nits, looks good. Haven't tried it though, would be good to add a test case.

**Comment by sotojuan** [comment]
The nipicks are fine, but not sure about testing because I don't think ANSI escape codes have been an issue or come up until the issue that led to this PR. That said, I have tested it locally and it works, but if anyone has a good idea of a test let me know.

**Comment by novemberborn** [comment]
> not sure about testing because I don't think ANSI escape codes have been an issue or come up until the issue that led to this PR

Sure, but it'd be good to have a test to protect against future regressions.

> if anyone has a good idea of a test let me know

Could add to [the TAP reporter tests](https://github.com/sindresorhus/ava/blob/master/test/reporters/tap.js), one "passing" with a title that contains escape codes, and one "failing".

**Comment by sindresorhus** [comment]
Hmm. Thinking out loud. Are we attacking this from the wrong side? Maybe we should color the test title in the reporters that needs it, not remove it here?

**Comment by sotojuan** [comment]
@novemberborn The ANSI codes show up when you test more than one file—is there a way to simulate that on the TAP tests?

**Comment by novemberborn** [comment]
> The ANSI codes show up when you test more than one file—is there a way to simulate that on the TAP tests?

@sotojuan I don't think that matters, the test calls `reporter.test()` directly.

> Are we attacking this from the wrong side? Maybe we should color the test title in the reporters that needs it, not remove it here?

Seems sensible to color when the test data is received, rather than in different reporters. Though admittedly there's more coloring going on in the reporters themselves. Maybe all of that can be improved in a future refactor?

**Comment by jamestalmage** [comment]
> Hmm. Thinking out loud. Are we attacking this from the wrong side? Maybe we should color the test title in the reporters that needs it, not remove it here?

Maybe, however, removing it here might be a good idea if TAP hates ansi codes in the test title.

I was just writing a test macro:

```js
function macro(obj, expected) {
  test(util.describe(obj, {colors: true}), t => {
    t.is(fn(obj), expected);
  });
}

macro({someObj}, 'foo');
macro({someOtherObj}, 'bar');
```

I thought the output with the verbose reporter looked awesome:

screenshot 2016-04-27 12 42 33

Maybe allowing this causes more headaches than it is worth, but I thought it was pretty cool!

**Comment by sotojuan** [comment]
I would say remove it just from TAP like I am doing now—seems like TAP stuff in general doesn't really like coloring in the raw reporter.

**Comment by sindresorhus** [comment]
@sotojuan Can you add `Fixes #00` (replace 00 with issue number) to the PR description so the relevant issue is closed when this is merged.

**Comment by sotojuan** [comment]
Thing is this was mentioned in issue (#723).

**Comment by sindresorhus** [comment]
Oh, right, then include just the issue number in the description. The more references the easier it is to navigate this at a later point.

**Comment by sindresorhus** [comment]
I don't think there's much benefit of testing the title alone. `stripAnsi` is already tested and all we do is to apply it. We should rather test the whole `tap` output to ensure there's no ANSI escape codes in any of the outputted information, not just the title.

https://github.com/chalk/has-ansi

**Comment by novemberborn** [comment]
> We should rather test the whole tap output to ensure there are no ANSI escape codes in any of the outputted information, not just the title.

👍

**Comment by sotojuan** [comment]
Added a test—let me know if it can be improved.

File: test/reporters/tap.js [hunk 3]
```diff
@@ -179,3 +180,18 @@ test('skip test', function (t) {
 	t.is(actualOutput, expectedOutput);
 	t.end();
 });
+
+test('reporter strips ANSI characters', function (t) {
+	var reporter = tapReporter();
+
+	var test = {
```
Comment (sindresorhus) [inline test/reporters/tap.js]: Just put this inline in the `reporter.test` call.

File: test/reporters/tap.js [hunk 4]
```diff
@@ -179,3 +180,18 @@ test('skip test', function (t) {
...
+		title: 'test \u001b[90m\u001b[2m›\u001b[22m\u001b[39m my test',
```
Comment (sindresorhus) [inline test/reporters/tap.js]: Use `chalk` to colorize it. Makes it easier if we ever want to change the style.

**Comment by sotojuan** [comment]
Done!

File: test/reporters/tap.js [hunk 5]
```diff
@@ -179,3 +181,16 @@ test('skip test', function (t) {
 	t.is(actualOutput, expectedOutput);
 	t.end();
 });
+
+test('reporter strips ANSI characters', function (t) {
+	var reporter = tapReporter();
+
+	var output = reporter.test({
+		title: chalk.gray.dim('test \u001b[90m\u001b[2m›\u001b[22m\u001b[39m my test'),
```
Comment (sindresorhus) [inline test/reporters/tap.js]: Should have been clearer. I meant instead of the ANSI escape codes.

```js
'test ' + chalk.gray.dim('›') + ' my test')
```

Comment (sotojuan) [inline test/reporters/tap.js]: Ha, no problem.

File: test/reporters/tap.js [hunk 6]
```diff
@@ -179,3 +181,16 @@ test('skip test', function (t) {
 	t.is(actualOutput, expectedOutput);
 	t.end();
 });
+
+test('reporter strips ANSI characters', function (t) {
+	var reporter = tapReporter();
+
+	var output = reporter.test({
+		title: chalk.gray.dim('test ' + chalk.gray.dim('›') + ' my test'),
```
Comment (sindresorhus) [inline test/reporters/tap.js]: The outer `chalk.gray.dim(` can be removed now.

Comment (sotojuan) [inline test/reporters/tap.js]: Ahhh, keep missing silly stuff, my bad.

**Comment by sotojuan** [comment]
All right, sorry about that (got a ton of things going on right now), should be fine now I hope 😎

**Comment by sindresorhus** [comment]
No worries. Looks good. 👍

Now go back to your studies. That's more important ;)

**Pull request merged**